File: archive/flutter-app/test/core/office_media/pdf_split_merge_test.dart
import 'dart:typed_data';
import 'dart:ui' show Rect, Size;

import 'package:flutter_test/flutter_test.dart';
import 'package:infrakit_studio/core/office_media/pdf_split_merge.dart';
import 'package:syncfusion_flutter_pdf/pdf.dart';

/// Builds a small, valid, in-memory PDF fixture with [pageCount] pages, each
/// carrying a distinguishing text label ("Page 1", "Page 2", ...) so tests
/// can tell pages apart without needing an external sample file — per the
/// task's guidance, `syncfusion_flutter_pdf` can both read and write PDFs,
/// so fixtures are built with its own `PdfDocument()` constructor rather
/// than checked-in binary files. [PdfDocument.save] is async, hence this
/// helper is too.
Future<Uint8List> _buildPdfAsync(int pageCount) async {
  final document = PdfDocument();
  for (var i = 1; i <= pageCount; i++) {
    final page = document.pages.add();
    page.graphics.drawString(
      'Page $i',
      PdfStandardFont(PdfFontFamily.helvetica, 12),
      brush: PdfSolidBrush(PdfColor(0, 0, 0)),
      bounds: const Rect.fromLTWH(0, 0, 150, 20),
    );
  }
  final saved = await document.save();
  document.dispose();
  return Uint8List.fromList(saved);
}

int _pageCountOf(Uint8List bytes) {
  final document = PdfDocument(inputBytes: bytes);
  final count = document.pages.count;
  document.dispose();
  return count;
}

void main() {
  group('PdfMerger', () {
    const merger = PdfMerger();

    test('merging two 1-page PDFs yields a 2-page PDF', () async {
      final a = await _buildPdfAsync(1);
      final b = await _buildPdfAsync(1);

      final merged = await merger.execute([a, b]);

      expect(_pageCountOf(merged), 2);
    });

    test('preserves page order and total count across three sources', () async {
      final a = await _buildPdfAsync(2);
      final b = await _buildPdfAsync(1);
      final c = await _buildPdfAsync(3);

      final merged = await merger.execute([a, b, c]);

      expect(_pageCountOf(merged), 6);
    });

    test('preserves each source page\'s own size instead of forcing a common size', () async {
      final document = PdfDocument();
      // Assign a fresh PdfPageSettings instance per page, with an explicit
      // orientation matching the size — see PdfMerger's doc comment for why
      // both matter (mutating `.size` is a no-op after the first page is
      // added, and omitting orientation lets Syncfusion's own internal
      // clone() swap width/height for landscape-shaped sizes).
      document.pageSettings = PdfPageSettings(const Size(300, 400), PdfPageOrientation.portrait);
      document.pages.add();
      document.pageSettings = PdfPageSettings(const Size(600, 200), PdfPageOrientation.landscape);
      document.pages.add();
      final custom = Uint8List.fromList(await document.save());
      document.dispose();

      final merged = await merger.execute([custom]);

      final reopened = PdfDocument(inputBytes: merged);
      expect(reopened.pages.count, 2);
      expect(reopened.pages[0].size.width, closeTo(300, 0.5));
      expect(reopened.pages[0].size.height, closeTo(400, 0.5));
      expect(reopened.pages[1].size.width, closeTo(600, 0.5));
      expect(reopened.pages[1].size.height, closeTo(200, 0.5));
      reopened.dispose();
    });

    test('produces a document that itself re-opens as a valid PDF', () async {
      final a = await _buildPdfAsync(1);
      final merged = await merger.execute([a]);

      // Re-opening and reading document-level state (not just page count)
      // is a reasonable proxy for "this is a well-formed PDF", since a
      // corrupt merge would typically throw here.
      final reopened = PdfDocument(inputBytes: merged);
      expect(reopened.pages.count, 1);
      reopened.dispose();
    });

    test('rejects an empty input list', () {
      expect(() => merger.execute(const []), throwsArgumentError);
    });
  });

  group('PdfPageExtractor', () {
    const extractor = PdfPageExtractor();

    test('extracting page 2 of a 3-page PDF yields a 1-page PDF', () async {
      final source = await _buildPdfAsync(3);

      final extracted = await extractor.execute(PdfExtractInput(bytes: source, pageNumbers: const [2]));

      expect(_pageCountOf(extracted), 1);
    });

    test('extracting multiple pages preserves ascending source order regardless of input order', () async {
      final source = await _buildPdfAsync(5);

      final extracted = await extractor.execute(PdfExtractInput(bytes: source, pageNumbers: const [4, 1, 2]));

      expect(_pageCountOf(extracted), 3);
    });

    test('duplicate page numbers are deduplicated', () async {
      final source = await _buildPdfAsync(3);

      final extracted = await extractor.execute(PdfExtractInput(bytes: source, pageNumbers: const [1, 1, 1]));

      expect(_pageCountOf(extracted), 1);
    });

    test('rejects an out-of-range page number', () async {
      final source = await _buildPdfAsync(2);

      expect(
        () => extractor.execute(PdfExtractInput(bytes: source, pageNumbers: const [5])),
        throwsArgumentError,
      );
    });

    test('rejects an empty page-number list', () async {
      final source = await _buildPdfAsync(2);

      expect(
        () => extractor.execute(PdfExtractInput(bytes: source, pageNumbers: const [])),
        throwsArgumentError,
      );
    });
  });

  group('PdfPageRangeParser', () {
    const parser = PdfPageRangeParser();

    test('parses a mix of single pages and ranges into ascending, deduplicated order', () {
      expect(parser.parse('1-3,5', pageCount: 5), [1, 2, 3, 5]);
    });

    test('deduplicates overlapping ranges and singles', () {
      expect(parser.parse('1-3,2,3-4', pageCount: 10), [1, 2, 3, 4]);
    });

    test('tolerates surrounding whitespace', () {
      expect(parser.parse(' 1 - 2 , 4 ', pageCount: 10), [1, 2, 4]);
    });

    test('rejects empty input', () {
      expect(() => parser.parse('', pageCount: 5), throwsFormatException);
    });

    test('rejects a malformed token', () {
      expect(() => parser.parse('1,two,3', pageCount: 5), throwsFormatException);
    });

    test('rejects a descending range', () {
      expect(() => parser.parse('5-2', pageCount: 5), throwsFormatException);
    });

    test('rejects a page number beyond the document\'s page count', () {
      expect(() => parser.parse('1-10', pageCount: 5), throwsRangeError);
    });

    test('rejects page number 0 (pages are 1-based)', () {
      expect(() => parser.parse('0-2', pageCount: 5), throwsRangeError);
    });
  });
}

File: app/src/core/office_media/pdfSplitMerge.test.ts
import { PDFDocument } from 'pdf-lib'
import { describe, expect, it } from 'vitest'
import { PdfMerger, PdfPageExtractor, PdfPageRangeParser } from './pdfSplitMerge'

/** Builds a tiny in-memory PDF with `pageCount` pages, each sized
 * `[width, height]` and (optionally) tagged with a `title`, so tests never
 * need on-disk fixture files. */
async function buildPdf(
  pageCount: number,
  options?: { width?: number; height?: number; title?: string },
): Promise<Uint8Array> {
  const doc = await PDFDocument.create()
  if (options?.title) doc.setTitle(options.title)
  for (let i = 0; i < pageCount; i++) {
    doc.addPage([options?.width ?? 200, options?.height ?? 300])
  }
  return doc.save()
}

describe('PdfMerger', () => {
  const merger = new PdfMerger()

  it('merges pages from multiple PDFs in source order', async () => {
    const a = await buildPdf(1)
    const b = await buildPdf(2)

    const mergedBytes = await merger.execute([a, b])
    const merged = await PDFDocument.load(mergedBytes)

    expect(merged.getPageCount()).toBe(3)
  })

  it('never reorders pages within a source document', async () => {
    const a = await buildPdf(2, { width: 100, height: 100 })
    const b = await buildPdf(1, { width: 400, height: 400 })

    const mergedBytes = await merger.execute([a, b])
    const merged = await PDFDocument.load(mergedBytes)

    expect(merged.getPageCount()).toBe(3)
    expect(merged.getPage(0).getWidth()).toBe(100)
    expect(merged.getPage(1).getWidth()).toBe(100)
    expect(merged.getPage(2).getWidth()).toBe(400)
  })

  it('preserves mixed page sizes and orientations across the merged set', async () => {
    const portrait = await buildPdf(1, { width: 200, height: 400 })
    const landscape = await buildPdf(1, { width: 400, height: 200 })

    const mergedBytes = await merger.execute([portrait, landscape])
    const merged = await PDFDocument.load(mergedBytes)

    expect(merged.getPage(0).getSize()).toEqual({ width: 200, height: 400 })
    expect(merged.getPage(1).getSize()).toEqual({ width: 400, height: 200 })
  })

  it('throws when given an empty list', async () => {
    await expect(merger.execute([])).rejects.toThrow('At least one PDF is required to merge')
  })

  it('throws when one of the source PDFs is empty', async () => {
    const a = await buildPdf(1)
    await expect(merger.execute([a, new Uint8Array(0)])).rejects.toThrow('One of the source PDFs is empty')
  })
})

describe('PdfPageExtractor', () => {
  const extractor = new PdfPageExtractor()

  it('extracts the requested 1-based pages', async () => {
    const source = await buildPdf(5, { width: 111, height: 222 })

    const outBytes = await extractor.execute({ bytes: source, pageNumbers: [2, 4] })
    const out = await PDFDocument.load(outBytes)

    expect(out.getPageCount()).toBe(2)
  })

  it('preserves source page order regardless of requested order', async () => {
    const doc = await PDFDocument.create()
    // Distinct sizes per page act as an identity fingerprint.
    doc.addPage([100, 100])
    doc.addPage([200, 200])
    doc.addPage([300, 300])
    const source = await doc.save()

    const outBytes = await extractor.execute({ bytes: source, pageNumbers: [3, 1] })
    const out = await PDFDocument.load(outBytes)

    expect(out.getPageCount()).toBe(2)
    expect(out.getPage(0).getWidth()).toBe(100)
    expect(out.getPage(1).getWidth()).toBe(300)
  })

  it('deduplicates repeated page numbers', async () => {
    const source = await buildPdf(3)
    const outBytes = await extractor.execute({ bytes: source, pageNumbers: [1, 1, 2] })
    const out = await PDFDocument.load(outBytes)
    expect(out.getPageCount()).toBe(2)
  })

  it('throws on a page number below 1', async () => {
    const source = await buildPdf(3)
    await expect(extractor.execute({ bytes: source, pageNumbers: [0] })).rejects.toThrow('out of range')
  })

  it('throws on a page number beyond the page count', async () => {
    const source = await buildPdf(3)
    await expect(extractor.execute({ bytes: source, pageNumbers: [4] })).rejects.toThrow(
      'the document has 3 page(s)',
    )
  })

  it('throws when given no page numbers', async () => {
    const source = await buildPdf(3)
    await expect(extractor.execute({ bytes: source, pageNumbers: [] })).rejects.toThrow(
      'At least one page number is required',
    )
  })

  it('throws when the source PDF is empty', async () => {
    await expect(extractor.execute({ bytes: new Uint8Array(0), pageNumbers: [1] })).rejects.toThrow(
      'Source PDF is empty',
    )
  })
})

describe('PdfPageRangeParser', () => {
  const parser = new PdfPageRangeParser()

  it('parses a mix of single pages and ranges into an ascending, deduplicated list', () => {
    expect(parser.parse('1-3,5,8-9', 10)).toEqual([1, 2, 3, 5, 8, 9])
  })

  it('deduplicates overlapping ranges and singles', () => {
    expect(parser.parse('1-3,2,3-4', 10)).toEqual([1, 2, 3, 4])
  })

  it('tolerates whitespace around tokens', () => {
    expect(parser.parse(' 1 - 3 , 5 ', 10)).toEqual([1, 2, 3, 5])
  })

  it('throws on empty input', () => {
    expect(() => parser.parse('', 10)).toThrow('Page range cannot be empty')
    expect(() => parser.parse('   ', 10)).toThrow('Page range cannot be empty')
  })

  it('throws on a non-numeric token', () => {
    expect(() => parser.parse('abc', 10)).toThrow('Invalid page number "abc"')
  })

  it('throws on a non-numeric token inside a range', () => {
    expect(() => parser.parse('1-abc', 10)).toThrow('Invalid range "1-abc"')
  })

  it('throws on a malformed range with too many hyphens', () => {
    expect(() => parser.parse('1-2-3', 10)).toThrow('Invalid range "1-2-3"')
  })

  it('throws when a range start is greater than its end', () => {
    expect(() => parser.parse('5-2', 10)).toThrow('start must be <= end')
  })

  it('throws when a page number is out of range for the document', () => {
    expect(() => parser.parse('1,15', 10)).toThrow('Page numbers must be between 1 and 10')
  })

  it('rejects trailing garbage on a numeric token (unlike parseInt)', () => {
    expect(() => parser.parse('5abc', 10)).toThrow('Invalid page number "5abc"')
  })
})
